src/swish/testing.ss

```diff
@@ -135,34 +136,64 @@
                          [`(DOWN ,@m ,@p ,r) ls])))))])
         (for-each
          (lambda (rogue)
-           (receive (after 100 (kill rogue 'shutdown))
+           (receive (after kill-delay (kill rogue 'shutdown))
              [`(DOWN ,m ,@rogue ,r) 'ok]))
          rogues))))
 
-  (define ($isolate-mat thunk)
-    (let* ([me self]
+  (define ($isolate-mat timeout deadline kill-delay thunk)
+    (let* ([white-flag
+            ;; avoid dynamic-wind frame in stack dump
+            (make-fault 'timeout)]
+           [me self]
            [pid (spawn
                  (lambda ()
                    (thunk)
                    (process-trap-exit #f)
                    (send me `#(done ,self))
                    (receive)))]
            [m (monitor pid)])
-      (on-exit (cleanup-after pid)
-        (receive (after 60000
+      (on-exit (cleanup-after pid deadline kill-delay)
+        (receive (after timeout
                    (kill pid 'kill)
-                   (throw 'timeout))
+                   (raise white-flag))
           [`(DOWN ,@m ,@pid ,reason ,e) (raise e)]
           [#(done ,@pid)
            (kill pid 'shutdown)
            (receive (after 1000 (throw 'timeout))
              [`(DOWN ,@m ,@pid shutdown) 'ok]
              [`(DOWN ,@m ,@pid ,r ,e) (raise e)])]))))
 
-  (define-syntax isolate-mat
-    (syntax-rules ()
-      [(_ name tags e1 e2 ...)
-       (mat name tags ($isolate-mat (lambda () e1 e2 ... (void))))]))
+  (define-syntax (isolate-mat x)
+    (define defaults
+      '([tags ()]
+        [timeout 60000]
+        [process-cleanup-deadline 500]
+        [process-kill-delay 100]))
+    (define (make-lookup x forms)
+      (define clauses
+        (collect-clauses x forms (map car defaults)))
+      (lambda (key)
+        (syntax-case (find-clause key clauses) ()
+          [(key val) #'val]
+          [_ (cond
+              [(assq key defaults) => cadr]
+              [else (errorf 'isolate-mat "unknown key ~s" key)])])))
+    (syntax-case x ()
+      [(_ name (isolate-mat-settings setting ...) e1 e2 ...)
+       (eq? (datum isolate-mat-settings) 'isolate-mat-settings)
+       (let ([lookup (make-lookup x #'(setting ...))])
+         (with-syntax ([tags (lookup 'tags)]
+                       [timeout-ms (lookup 'timeout)]
+                       [deadline (lookup 'process-cleanup-deadline)]
+                       [kill-delay (lookup 'process-kill-delay)])
+           #`(add-mat 'name 'tags
+               #,(replace-source x ;; use isolate-mat source in timeout continuation
+                   #'(lambda ()
+                       ($isolate-mat timeout-ms deadline kill-delay
+                         (lambda () e1 e2 ... (void)))
+                       (void))))))]
+      [(_ name tag-list e1 e2 ...)
+       #'(isolate-mat name (isolate-mat-settings [tags tag-list]) e1 e2 ...)]))
```

Comment:
Will this case also preserve the source information?

Reply:
Excellent question. In my testing the expected source information is preserved in both cases, though it's not immediately obvious why it works.

I half expected we might need to replace `_` on the original line 195 with a pattern variable and reference than in place of `isolate-mat` on 196. That may be more reliable or at least easier to fathom.

Another option is to use `#'name` instead of `x` as the source for `replace-source`, but that may mislead someone looking at the stack to think that mat names are bound to procedures.